cpp/Naive.cpp

```diff
@@ -1,65 +1,413 @@
-        X = initialMoments.extent(1),
-        Y = initialMoments.extent(2);
-        assert(X == Y);
 
-        fftw::mdbuffer<3u> moments{M, X, Y};
-        for (int m = 0; m < M; ++m) {
-            for (int x = 0; x < X; ++x) {
-                for (int y = 0; y < Y; ++y) {
-                    moments(m, x, y) = initialMoments(m, x, y);
-                }
-            }
-        }
+    template<class T, size_t Size, class ...Args>
+    std::array<T, Size> forward_to_array(Args &&...args) {
+        return forward_to_array_impl<T>(std::make_index_sequence<Size>(), std::forward<Args>(args)...);
+    }
+}
+
+namespace ahr {
+    Naive::Naive(std::ostream &out, Dim M, Dim X, Dim Y) : HermiteRunner(out), M(M), X(X), Y(Y) {}
+
+    void Naive::init(Dim N_) {
+        N = N_;
+
+        // Currently assuming X==Y for simplicity, but the code is written generally for the most part.
+        assert(X == Y);
 
-        auto inSubspan = getMomentSlice(moments, 0);
-        auto outSubspan = getMomentSlice(momentsPH, 0);
+        // Plan FFTs both ways
+        fft = fftw::plan<2u>::dft(phi_K, phi_K_New, fftw::FORWARD, fftw::MEASURE);
+        fftInv = fftw::plan<2u>::dft(phi_K_New, phi_K, fftw::BACKWARD, fftw::MEASURE);
 
-        plan = fftw::plan<2u>::dft(inSubspan, outSubspan, fftw::FORWARD, fftw::MEASURE);
+        // Initialize equilibrium values
+        auto [aParEq, phi] = equilibriumOT01(X, Y);
+        fft(phi, phi_K);
+        fft(aParEq, aParEq_K);
 
+        // DEBUG
+        print(aParEq);
+        print(aParEq_K);
+        // Transform moments into phase space
         for (int m = 0; m < M; ++m) {
-            plan(getMomentSlice(moments, m), getMomentSlice(momentsPH, m));
+            if (m == A_PAR) continue;
+            for_each_xy([&](Dim x, Dim y) {
+                moments_K(m, x, y) = 0;
+            });
         }
+
+        // aPar and uekPar
+        for_each_kxky([&](Dim ky, Dim kx) {
+            aParEq_K(kx, ky) /= std::sqrt(KX*KY);
+            moments_K(A_PAR, kx, ky) = aParEq_K(kx, ky);
+            ueKPar_K(kx, ky) = kPerp(kx, ky) * moments_K(A_PAR, kx, ky);
+        });
+        print(aParEq_K);
     }
 
     void Naive::run() {
+        // store all derivatives
+        DxDy<fftw::mdbuffer<3u>> dGM{M, X, Y};
+        DxDy<fftw::mdbuffer<2u>> dPhi{X, Y}, dUEKPar{X, Y};
+
+        bool divergent = false, repeat = false, noInc = false;
+        Real dt{0};
+        HyperCoefficients hyper{};
 
         for (int t = 0; t < N; ++t) {
             // predictor step
+            derivatives(phi_K, dPhi);
+            derivatives(ueKPar_K, dPhi);
             for (int m = 0; m < M; ++m) {
-// TODO
-//                plan(getMomentSlice(m), getMomentSlice());
+                derivatives(sliceXY(moments_K, m), sliceXY(dGM, m));
+            }
+
+            // DEBUG
+            print(sliceXY(moments_K, A_PAR));
+
+            if (repeat or divergent) {
+                t--; // repeat previous timestep
```

Comment:
Note that the difference between repeat and divergent is that the noinc bool is set to true for repeat. This affects some stuff in timestep calcs later, but shouldn't be code-breaking